compiler/rustc_borrowck/src/type_check/mod.rs

```diff
@@ -1270,14 +1270,17 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
                 if let Err(terr) =
                     self.sub_types(rv_ty, place_ty, location.to_locations(), category)
                 {
-                    span_mirbug!(
-                        self,
-                        stmt,
-                        "bad assignment ({:?} = {:?}): {:?}",
-                        place_ty,
-                        rv_ty,
-                        terr
-                    );
+                    // HACK(ouz-a): Bypasses ICE for #91633
+                    if rv_ty.discriminant_ty(self.tcx()) != place_ty.discriminant_ty(self.tcx()) {
```

Comment:
I'm not certain that this is the right check. `Ty::discriminant_ty` seems like kind of random here. For example, this also allows the bad assignment `ty::Bool <- ty::Char`, since both of them have the same discriminant: https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_middle/ty/sty.rs.html#2046-2047

Reply:
Ahh, good catch!

Reply:
I'm not exactly sure what the _right_ solution is here -- 

I think we could fix this by avoiding lowering `&it[0]` to the [`Index`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/thir/enum.ExprKind.html#variant.Index) operator in THIR if the param-env has a predicate that satisfies `[T]: std::ops::Index<_>`. That would avoid having to touch how we normalize projection types...

Reply:
yeah, imo the right solution should be to do candidate selection and check whether we end up with a lang item impl (or an impl on the lang item as the impls are currently not tagged as lang items iirc)

Reply:
probably sufficient to check that the selection candidate is an `ImplCandidate` and that's it (b/c the only impl candidate that could apply to `[_]: Index<usize>` is the built-in one, due to coherence?)

Reply:
So we actually can't do it here during THIR build, because we remove the type_dependent_def during typeck writeback if we think it's a builtin impl, so trying to lower it as a THIR method call ICEs.

This new solution @ouz-a implemented in writeback makes a lot more sense.